draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1299 @@
+```
+
+Let me reformat this line to make clearer what's happening here:
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+Starting from the inside - this:
+
+```Ruby
+  call(arguments)
+```
+
+translates to:
+
+```Ruby
+CLI.new(arguments).config
+```
+
+This is the part that `Procto` provides.
+
+So this
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+means:
+
+* we're getting back a configuration based on our CLI arguments
+* then pass that to `Env::Bootstrap.call`
+* pass whatever comes back from that to our Runner
+
+What does the configuration look like?
+
+### Configuration
+
+If you check out the constructor of `Mutant::CLI` you'll see this:
+
+```Ruby
+def initialize(arguments)
+  @config = Config::DEFAULT
+
+  parse(arguments)
+end
+```
+
+Ok, nice and easy. First we'll set the configuration to our defaults, then we're parsing the cli arguments and overwrite our configuration based on this.
+
+What does the default configuration look like?
+For that you'll have to check out the top level library file, so lib/`mutant.rb`:
+
+
+```Ruby
+module Mutant
+  # Reopen class to initialize constant to avoid dep circle
+  class Config
+    DEFAULT = new(
+      debug:             false,
+      expected_coverage: Rational(1),
+      # snip
+      fail_fast:         false,
+      includes:          EMPTY_ARRAY,
+      # snip
+      kernel:            Kernel,
+      load_path:         $LOAD_PATH,
+      matcher:           Matcher::Config::DEFAULT,
+      open3:             Open3,
+      pathname:          Pathname,
+      requires:          EMPTY_ARRAY,
+      reporter:          Reporter::CLI.build($stdout),
+      zombie:            false
+    )
+  end # Config
+end # Mutant
+```
+
+I did remove a lot of the config object above to keep it readable. I just want to highlight a couple of lines I find interesting:
+
+```Ruby
+expected_coverage: Rational(1)
+```
+
+This means that the expected default coverage is 100%. Yep, 100%. You can overwrite this though. Suppose your `Mutant` run fails with this result:
+
+```
+Mutations: 197
+Kills: 178
+```
+
+You can make it pass like this:
+
+```
+bundle exec mutant --expected-coverage 179/197 # plus the rest of your cli arguments
+```
+
+Back to the config:
+
+```Ruby
+load_path:         $LOAD_PATH,
+requires:          EMPTY_ARRAY,
+pathname:          Pathname,
+```
+
+What I like here is how configurable `Mutant`. You can even inject another $LOAD_PATH, additional requires or another Pathname'esque library.
+
+But the configurability is actually just a sideeffect here. The main goal is to stop relying on IO globals read from a global scope and injecting them explicitly.
+
+This lists the IO dependencies of each unit much more explicit. Makes testing easier, and allows to write a testing tool that works without message expectations (one target of mine in the near future).
```

Comment:
Yep, stupid typo / c&p error on my side, probably a couple more of those further down there.
Proper proof reading is still on my todo list :)